### PR TITLE
JAMES-3924 RabbitMQ mail queue updates: not performed

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 1]
```diff
@@ -700,7 +731,9 @@ private void dequeueMails(int times) {
                 .subscribe();
 
             try {
-                await().untilAsserted(() -> assertThat(counter.get()).isGreaterThanOrEqualTo(times));
+                await()
+                        .atMost(Duration.ofMinutes(10))
+                        .untilAsserted(() -> assertThat(counter.get()).isGreaterThanOrEqualTo(times));
```
Comment (Arsnael) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:736]: indent?

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 2]
```diff
@@ -678,9 +705,13 @@ void messagesShouldBeProcessedAfterNotPublishedMailsHaveBeenReprocessedAndNewMes
 
         private void enqueueSomeMails(Function<Integer, String> namePattern, int emailCount) {
             IntStream.rangeClosed(1, emailCount)
-                .forEach(Throwing.intConsumer(i -> enQueue(defaultMail()
-                    .name(namePattern.apply(i))
-                    .build())));
+                .forEach(Throwing.intConsumer(i -> {
+                    FakeMail mail = defaultMail()
+                            .name(namePattern.apply(i))
+                            .build();
```
Comment (Arsnael) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:711]: indent?

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 3]
```diff
@@ -204,8 +206,9 @@ void browseShouldReturnCurrentlyEnqueuedMailFromAllSlices() throws Exception {
         void browseStartShouldBeUpdated(CassandraCluster cassandraCluster) {
             int emailCount = 100;
 
-            StatementRecorder statementRecorder = new StatementRecorder();
-            cassandraCluster.getConf().recordStatements(statementRecorder);
+            StatementRecorder.Selector selector = preparedStatementStartingWith("UPDATE browsestart");
+            StatementRecorder statementRecorder = cassandraCluster.getConf()
+                    .recordStatements(selector);
```
Comment (Arsnael) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:211]: indent?

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1623/4/testReport/

Related?

**Comment by chibenwa** [comment]
[org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$DeDuplicationTest.dequeueShouldStillRetrieveAllBlobsWhenIdenticalContentAndDeduplication](https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1623/4/testReport/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$DeDuplicationTest/dequeueShouldStillRetrieveAllBlobsWhenIdenticalContentAndDeduplication/)

possibly, yes.

**Comment by chibenwa** [comment]
> Here when we dequeue you trying to dispose of the mail item right away. So in the test of course when you trying to access the content of it, it returns a NPE. Seeing as well how the mail item is trying to be delivered after being dequeued in theJamesMailSpooler for example, I'm not sure we want to do a dispose here. I think we should revert this commit, but maybe I misunderstood sth?

Or add a check to avoid the NPE...

**Comment by Arsnael** [comment]
> Or add a check to avoid the NPE...

That maybe avoids the NPE but I don't see how it's fixing the test? Also you didn't really answer my questions and concerns on why you want to dispose the mail object here? Your test is trying to check with dedup that if we dequeue two identical mails (subject + content) we check we dequeue two emails, and checking they have the same mime message content... But because of the dispose the content has been disposed of. Seeing also as we deliver those mails after dequeing them, does it mean we send them without content then?

As I said above, I'm not sure I understood well that part of the code, so a bit more detailed answer would be appreciated.

**Comment by quantranhong1999** [comment]
I do not have permission to push the test fix commit to Benoit's branch.

Please cherry-pick or add me push permission.
https://github.com/quantranhong1999/james-project/commits/JAMES-3924

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Still failing apparently...



 
--
 

<a href="https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/">Test Result</a> (3 failures  / +2)<ul style="list-style-type: none; margin: 0;"><li style="" class="shown"><a href="https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/enqueuedEmailsShouldEventuallyBeCleaned/">org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.enqueuedEmailsShouldEventuallyBeCleaned</a></li><li style="" class="shown"><a href="https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/browseStartShouldBeUpdated_CassandraCluster_/">org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.browseStartShouldBeUpdated{CassandraCluster}</a></li><li style="" class="shown"><a href="https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/contentStartShouldBeUpdated_CassandraCluster_/">org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.contentStartShouldBeUpdated{CassandraCluster}</a></li></ul>[Test Result](https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/) (3 failures / +2)

    [org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.enqueuedEmailsShouldEventuallyBeCleaned](https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/enqueuedEmailsShouldEventuallyBeCleaned/)
    [org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.browseStartShouldBeUpdated{CassandraCluster}](https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/browseStartShouldBeUpdated_CassandraCluster_/)
    [org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.contentStartShouldBeUpdated{CassandraCluster}](https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/contentStartShouldBeUpdated_CassandraCluster_/)

**Comment by Arsnael** [comment]
> Still failing apparently...

=> seems to join my observations in https://github.com/apache/james-project/pull/1623#issuecomment-1630376345 , second bullet point. I could notice tests in there being unstable even locally, smetimes green and sometimes red

**Comment by chibenwa** [comment]
Relaunching the build and crossing fingers then...

**Comment by Arsnael** [comment]
It feels like those changes make the tests unstable, I think it was more or less stable before that. Shouldn't we check a bit deeper?

**Comment by chibenwa** [comment]
> It feels like those changes make the tests unstable, I think it was more or less stable before that. Shouldn't we check a bit deeper?

Of course!

**Comment by chibenwa** [comment]
BTW [org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.browseStartShouldBeUpdated{CassandraCluster}](https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/browseStartShouldBeUpdated_CassandraCluster_/)
[org.apache.james.queue.rabbitmq.RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled.contentStartShouldBeUpdated{CassandraCluster}](https://ci-builds.apache.org/job/james/job/ApacheJames/view/change-requests/job/PR-1623/9/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/contentStartShouldBeUpdated_CassandraCluster_/)

Are introduced in this PR

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 4]
```diff
@@ -107,7 +110,7 @@
 class RabbitMQMailQueueTest {
     private static final HashBlobId.Factory BLOB_ID_FACTORY = new HashBlobId.Factory();
     private static final int THREE_BUCKET_COUNT = 3;
-    private static final int UPDATE_BROWSE_START_PACE = 2;
+    private static final int UPDATE_BROWSE_START_PACE = 10;
```
Comment (chibenwa) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:113]: Increase this a bit more?

```suggestion
    private static final int UPDATE_BROWSE_START_PACE = 20;
```

**Comment by Arsnael** [comment]
Still unstable unfortunately: https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1623/15/testReport/

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 5]
```diff
@@ -197,6 +202,55 @@ void browseShouldReturnCurrentlyEnqueuedMailFromAllSlices() throws Exception {
...
+            dequeueMails(emailCount);
+
+            clock.setInstant(IN_SLICE_3);
+            enqueueSomeMails(namePatternForSlice(3), emailCount);
+            dequeueMails(emailCount);
+
+            // The actual rate of update should actually be lower than the update probability.
+            assertThat(statementRecorder.listExecutedStatements(selector))
+                .hasSizeBetween(2, 12);
+        }
```
Comment (quantranhong1999) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:228]: Basically, this test asserts a certain statement on something that depends on probability. It means it always has a chance to fail.
We even can calculate (theoretically) the failure percentage of this test using the Bernoulli trial, and the result is not small.
I suggest we mark these tests as `@Unstable`. WDYT @Arsnael @chibenwa ?

Comment (Arsnael) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:228]: I'm not sure. I understand your comment and it makes sense. But this test seemed pretty stable to me before this change, now it's very unstable. 

Also it feels different things are being performed in this PR, could potentially cut it smaller?

Comment (quantranhong1999) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:228]: > But this test seemed pretty stable to me before this change, now it's very unstable.

No.
https://github.com/apache/james-project/pull/1623#issuecomment-1639303055

Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:228]: > We even can calculate (theoretically) the failure percentage of this test using the Bernoulli trial, and the result is not small.

That's the point: the test is here to ensure we avoid updating the browse start when not needed.

Upper value verifies that if current browseStart is still in the current slice we do not update. BrowseStart updates count should thus be '3' or '4' but exactly the count of slice switches.

Now the rabbitMQ prefetch fucks this up by introducing concurrency issues. Setting the QOS to 1 solved the issue. We can't set this up for the all test suite but using a QOS of 1 in this test would INDEED solve our issue.

Cc @quantranhong1999

Comment (quantranhong1999) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java:228]: Thank you for your suggestion. It indeed helps the tests to be more stable.

**Comment by chibenwa** [comment]
See https://github.com/apache/james-project/pull/1660

**Pull request closed without merging**